### PR TITLE
Change visibility to pub for all structs and field

### PR DESCRIPTION
**Pull request by JuliDi**
This PR aims to fix issue #2 by moving the fields' `pub` specifier inside the parentheses.

Further, the visibility specifier and other macros (mainly docstring in case you have a `#[deny(missing_docs)]` in your crate) is parsed, so you can now do

```rust 
assign_resources::assign_resources! {
    /// Test Struct
    monitoring_adc: pub MonitoringAdcResources {
        /// Test Field1
        spi: SPI6,
        /// Test Field2
        sck: PA5, //...
```
and get
```rust
#[allow(dead_code, non_snake_case)]
#[doc = " Test Struct"]
pub struct MonitoringAdcResources {
    #[doc = " Test Field1"]
    pub spi: peripherals::SPI6,
    #[doc = " Test Field2"]
    pub sck: peripherals::PA5,
    //....
}
```

As mentioned in the issue, I am far from an expert when it comes to Rust macros, so any feedback is hightly appreciated!

### CODE REVIEW

**Comment by JuliDi** [comment]
Struggling with how to get the `split_resources` macro to be public along with the `_ResourceAssigs`. Seems like you cannot just use `$crate::_ResourceAssigs`

**Comment by JuliDi** [comment]
With the latest commit, I can now have the resources assigned in a file `src/resources.rs` and then do a `use my_lib::resources::*`.

This is not yet ideal, I would like to only need to `use` the `assign_resources` macro and then everything is usable, as well. But maybe that's not even possible.

**Comment by adamgreig** [comment]
Thanks for the report and PR! Maybe we could have one PR just to fix #2 and leave the other changes to a separate PR?

I wonder if it would be worth making the visibility of both structs and fields user-controlled, i.e. allow `pub` before each and copy it into the output. I think most users are either calling this in the final binary firmware, in which case being pub or not makes no difference, or in another crate that exports it, in which case it needs to be pub, so perhaps pub-by-default is best.

**Comment by JuliDi** [comment]
> Thanks for the report and PR! Maybe we could have one PR just to fix #2 and leave the other changes to a separate PR?

Absolutely, I just kept editing by branch to make it work – which go a bit out of hand.  I will create a new one for only #2 .

I agree that pub by default is best. It does no harm if it is there and avoids a lot of trouble.
Regarding the `use xyz::*` statements: Any idea if/how that could be solved?

**Comment by adamgreig** [comment]
>  Regarding the use xyz::* statements: Any idea if/how that could be solved?

I'm not sure it can. `$crate` refers to the crate that defined the macro, so in this case I think it should be the crate that contains the call to `assign_resources()`, so I would have thought you could use it to refer to the struct... but 1) perhaps it's actually going to refer to the `assign-resources` crate, since it's used inside the expansion of the `assign_resources!()` macro, and 2) even so, you don't know what path into the crate the macro was called at, so you still don't know what path to import the struct from.

Perhaps the best we can do is make the generated top-level struct a sensible name like `AssignResources`, make it pub, and the end-user has to make sure it's in scope when they call `split_resources`, along with the `split_resources` macro itself and presumably any of the sub-structs. We _could_ allow the user to define the struct name with something like

```rust
assign_resources! {
    AssignResources {
        led: LedResources {
            green: PA1,
       }
    }
}
```

but I don't love the extra indentation and it becomes a slightly weird syntax too.

Maybe a better alternative for people wanting to put the resource definitions into a library or module is for them to define their own macro which calls the assign_resources macro, and then call their macro in their top-level application. That way it's just as though they'd written out the assignments at the top level, everything is in scope, but you can still have the nitty-gritty in the library module and use cfg/different dependencies to change it for different boards/targets/etc.

**Comment by JuliDi** [comment]
Sure, this should not become too complicated. Making everything pub and documenting how to use it in a library (i.e., what to import) would suffice from my point of view.
It's not a huge macro, so if you need full control you can just write a new one.

How about making the structs and fields pub, renaming the one struct to AssignResources (and documenting that) and then adding a small example for how to use it in a library crate?

That would completely cover my use case and would not require a lot of changes.

**Comment by adamgreig** [comment]
Sounds fine to me. By defining your own macro, I mean more like:

`lib/resources.rs`:
```rust
macro_rules! resources {
() => {
  assign_resources::assign_resources! {
    led: LedResources {
      red: PA1,
      green: PA2,
    },
    usb: UsbResources {
      usb: USBD,
      dma: DMA1_CH2,
    },
  }
}}
```

`app/main.rs`:
```rust
use lib::resources::resources;
resources!();
fn leds(leds: LedResources) { ... }
fn main() { p = ...; r = split_resources!(p); }
```

in other words, the library defines a macro that just calls the `assign_resources` macro, then `main` only needs to import that macro, not any structs or anything else, and they get created when it's called. Might work if you only need to refer to those structs inside the application and just want the definition elsewhere (e.g. to make it easier to swap for another board), but not helpful if the library itself also needs the resource types.

Anyway, yes, I think let's make the top-level struct `AssignedResources`, document the name, and then both ways of using it are fine and fairly easy.

**Comment by JuliDi** [comment]
I've made some changes and added documentation. Also pushed the code I used for testing in case you want to verify that everything is working: https://github.com/JuliDi/assign-resources-test

**Comment by adamgreig** [comment]
Thanks, looks good! Couple of last points...

* Why are the outer and inner captures different (ie meta vs ident+args)?
* We should specifically document that the generated struct is called `AssignedResources` in the docstring
* Could you rebase to resolve the merge conflict?
* If we move the type aliases to after the new struct, the user can put a docstring above the macro call site to document the generated AssignedResources struct I believe, so then it might not need allow(missing_docs).

**Comment by JuliDi** [comment]
Thanks for your review! I will look into it in the next couple of days and address all your points.

Regarding the meta vs ident+args: if I recall correctly, something was not working the way I wanted and it may just be a copy&paste artifact. So no particular reason...

**Comment by JuliDi** [comment]
The docstrings are a bit nasty. For one, rust analyzer fails to properly expand the macro when I add docstrings to the struct fields (rust rover, which iirc uses caro expand or some other crates.io tool, manages to do so). 
Further, your idea of placing the docstring above the macro call would (in theory) produce exactly what we want. However, it is not recognized as a docstring by the compiler, which causes errors when `deny(missing_docs)` is set.

From this warning, I conclude that it is not possible to do achieve what we're trying:
```
warning: unused doc comment
  --> src/main.rs:13:1
   |
13 | /// Resource Assignment
   | ^^^^^^^^^^^^^^^^^^^^^^^ rustdoc does not generate documentation for macro invocations
   |
   = help: to document an item produced by a macro, the macro must produce the documentation as part of its expansion
```
I've rebased the branch so all merge conflicts are resolved now.

File: src/lib.rs [hunk 1]
```diff
@@ -53,28 +57,38 @@
...
+            $(#[$outer])*
+            pub struct $group_struct {
+                $(
+                    $(#[$inner])*
+                    pub $resource_name: peripherals::$resource_field
+                ),*
             }
         )+
+
+        $($($(type $resource_alias = peripherals::$resource_field;)?)*)*
```
Comment (adamgreig) [inline src/lib.rs]:
```suggestion
        $($($(pub type $resource_alias = peripherals::$resource_field;)?)*)*
```

**Comment by adamgreig** [comment]
Thanks! I guess the only other option is to syntactically move the outer struct to inside the macro, so the user could set the name and any documentation. I worry that would look a bit annoying for most users, but maybe it's nice to let the name be configurable.

I can't think of any nice syntax for that off-hand, so if you have any good ideas I'm open to them, otherwise let's get this merged (plus perhaps make the type alias `pub` above?) and published.

**Comment by JuliDi** [comment]
Unfortunately I have no good idea either. But for now I think we don't need to overengineer this.
Feel free to merge it :)

**Comment by adamgreig** [comment]
thanks, published!

**Pull request merged**